ocaml-lsp-server/src/semantic_highlighting.ml

```diff
@@ -10,142 +10,80 @@ open Fiber.O
 
 module Token_type : sig
-  type t = private int
-
-  val module_ : t
-
-  val type_ : t
+  type t
 
-  val class_ : t
+  val of_builtin : SemanticTokenTypes.t -> t
 
-  val enum : t
+  val module_ : t
 
   val module_type : t
 
-  val record : t
-
-  val type_parameter : t
-
-  val parameter : t
-
-  val variable : t
-
-  val property : t
-
-  val enum_member : t
-
-  val event : t
-
-  val function_ : t
-
-  val method_ : t
-
-  val macro : t
-
-  val keyword : t
-
-  val modifier : t
-
-  val comment : t
-
-  val string : t
-
-  val number : t
-
-  val regexp : t
-
-  val operator : t
-
-  val decorator : t
-
-  val list : string list
+  val to_int : t -> int
 
   val to_legend : t -> string
-end [@warning "-32"] = struct
-  type t = int
-
-  let module_ = 0
-  (* the number is determined by the index in [list] defined below; e.g., we use
-     [0] here because we'd like to use "namespace" token type for module
-     symbols *)
-
-  let type_ = 1
-
-  let class_ = 2
-
-  let enum = 3
-
-  let module_type = 4
-
-  let record = 5
-
-  let type_parameter = 6
-
-  let parameter = 7
-
-  let variable = 8
-
-  let property = 9
-
-  let enum_member = 10
 
-  let event = 11
-
-  let function_ = 12
-
-  let method_ = 13
-
-  let macro = 14
-
-  let keyword = 15
-
-  let modifier = 16
-
-  let comment = 17
-
-  let string = 18
-
-  let number = 19
-
-  let regexp = 20
-
-  let operator = 21
-
-  let decorator = 22
-
-  let list =
-    [ "namespace"
-    ; "type"
-    ; "class"
-    ; "enum"
-    ; "interface"
-    ; "struct"
-    ; "typeParameter"
-    ; "parameter"
-    ; "variable"
-    ; "property"
-    ; "enumMember"
-    ; "event"
-    ; "function"
-    ; "method"
-    ; "macro"
-    ; "keyword"
-    ; "modifier"
-    ; "comment"
-    ; "string"
-    ; "number"
-    ; "regexp"
-    ; "operator"
-    ; "decorator"
+  val tokenTypes : string list
+end = struct
+  type t = SemanticTokenTypes.t
+
+  let of_builtin t = t
+
+  let module_ = SemanticTokenTypes.Namespace
+
+  let module_type = SemanticTokenTypes.Interface
+
+  let legend : SemanticTokenTypes.t list =
+    [ Namespace
+    ; Type
+    ; Class
+    ; Enum
+    ; Interface
+    ; Struct
+    ; TypeParameter
+    ; Parameter
+    ; Variable
+    ; Property
+    ; EnumMember
+    ; Event
+    ; Function
+    ; Method
+    ; Macro
+    ; Keyword
+    ; Modifier
+    ; Comment
+    ; String
+    ; Number
+    ; Regexp
+    ; Operator
+    ; Decorator
     ]
 
-  let array = lazy (Array.of_list list)
+  let tokenTypes : string list =
+    List.map legend ~f:(fun s ->
+        match SemanticTokenTypes.yojson_of_t s with
+        | `String s -> s
+        | _ -> assert false)
+
+  let to_int =
+    let module Table = MoreLabels.Hashtbl in
+    let table =
+      lazy
+        (let t = Table.create (List.length legend) in
+         List.iteri legend ~f:(fun data key -> Table.add t ~key ~data);
+         t)
+    in
+    fun t -> Table.find (Lazy.force table) t
```

Comment:
we could replace this with a pattern match - should be much more efficient without hashing? (although I don't know how ocaml hashes constructors - just integers essentially - may be a cheap thing) (again would introduce magic ints, however)

Reply:
The ints aren't the problem, it's just that this large pattern match is going to add a lot of code for performance gains that are very likely to be superficial.

Reply:
Sounds fair

Reply:
I'd suggest we focus on optimizing once we have a nice and general purpose API for semantic tokens in the lsp library.